core/src/main/java/org/testcontainers/containers/ComposeServiceWaitStrategyTarget.java

```diff
@@ -0,0 +1,68 @@
+package org.testcontainers.containers;
+
+import com.github.dockerjava.api.command.InspectContainerResponse;
+import com.github.dockerjava.api.model.Container;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import org.slf4j.Logger;
+import org.testcontainers.DockerClientFactory;
+import org.testcontainers.containers.wait.strategy.WaitStrategyTarget;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+
+/**
+ * Class to provide a wait strategy target for services started through docker-compose
+ */
+@EqualsAndHashCode
+class ComposeServiceWaitStrategyTarget implements WaitStrategyTarget {
+
+    private final Container container;
+    private final GenericContainer proxyContainer;
+    @Getter
+    private final Logger logger;
+    private Map<Integer, Integer> mappedPorts = new HashMap<>();
+    @Getter
+    private List<Integer> exposedPorts = new ArrayList<>();
+    private InspectContainerResponse containerInfo;
+
+    ComposeServiceWaitStrategyTarget(Container container, GenericContainer proxyContainer,
+                                     Logger logger, Map<Integer, Integer> mappedPorts) {
+        this.container = container;
+
+        this.proxyContainer = proxyContainer;
+        this.logger = logger;
+
+        if (mappedPorts != null) {
```

Comment:
A `ComposeServiceWaitStrategyTarget` is created for every service returned by `listChildContainers()` in `DockerComposeContainer`, which can return services with no exposed ports.

for example:
```yaml
redis:
  image: redis
db:
  image: orchardup/mysql
```

and a test with
```java
new DockerComposeContainer(new File("src/test/resources/compose-test.yml"))
            .withExposedService("redis_1", REDIS_PORT, Wait.forListeningPort());
```

Only the redis container has exposed/mapped ports, so the `ComposeServiceWaitStrategyTarget` for the `db` service would have a null value for `mappedPorts`.

I'll change the behaviour to only create a `ComposeServiceWaitStrategyTarget` for services that have been explicitly exposed.

Reply:
We should never accept null as a valid collection argument's value, that was the motivation :)

Reply:
> I'll change the behaviour to only create a ComposeServiceWaitStrategyTarget for services that have been explicitly exposed.

Now I think about use cases where waiting strategy doesn't use the ports at all but checks the logs for instance.
Current API doesn't allow that without exposing a port, right?

Should we maybe add `waitingFor("redis_1", Wait.forLogMessage())` method?

Reply:
I can add that method, but then we get into the problem of a `ComposeServiceWaitStrategyTarget` with no exposed ports, so the constructor would have to accept a null or empty value for `mappedPorts`, which would mean returning a empty list from `getExposedPorts`.

That goes against this idea:

> We should never accept null as a valid collection argument's value, that was the motivation :)

It also means you could use `DockerComposeContainer` like this:

```java
new DockerComposeContainer(new File("src/test/resources/compose-test.yml"))
            .withExposedService("redis_1", REDIS_PORT, Wait.forListeningPort());
```

or like this:

```java
new DockerComposeContainer(new File("src/test/resources/compose-test.yml"))
            .withExposedService("redis_1", REDIS_PORT)
            .waitingFor("redis_1", Wait.forListeningPort());
```

Is that slightly confusing for the user?

Reply:
>  but then we get into the problem of a ComposeServiceWaitStrategyTarget with no exposed ports, so the constructor would have to accept a null or empty value for mappedPorts, which would mean returning a empty list from getExposedPorts

Sounds totally fine to me - no exposed ports means an empty set.

Second example IMO doesn't look confusing, and makes even more sense because we're actually waiting for a service and not for the exposed port :)

Just `.withExposedService("redis_1", REDIS_PORT, Wait.forListeningPort())` is a good alias/shortcut for it